libc/stdio_printf.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libc-tests
+ *
+ * libphoenix fprintf, snprintf, sprintf formatting tests
+ * 
+ * Copyright 2021 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <stddef.h>
+
+#include <unity_fixture.h>
+
+
+FILE *wr, *rd;
+char buf[50];
+char stdpath[] = "stdio_printf_test";
+int i;
+
+/* comma delimiter macro ',' for multiple arguments parsing to *printf() function via macros */
+#define COMMA ,
+
+/* macro-assertion of fprintf formatting */
+#define ftest(_format, _value, _expect) \
+	{ \
+		rewind(wr); \
+		rewind(rd); \
+		TEST_ASSERT_EQUAL_INT(strlen(_expect), fprintf(wr, _format, _value)); \
+		fputc('\n', wr); \
+		fflush(wr); \
+		fgets(buf, strlen(_expect) + 1, rd); \
+		TEST_ASSERT_EQUAL_STRING(_expect, buf); \
+	}
+
+/* macro-assertion of sprintf formatting */
+#define stest(_format, _value, _expect) \
+	{ \
+		memset(buf, 0, 50); \
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, sprintf(buf, _format, _value)); \
+		TEST_ASSERT_EQUAL_STRING(_expect, buf); \
+	}
+
+#define sntest(_snret, _snlen, _format, _value, _expect) \
+	{ \
+		memset(buf, 0, 50); \
+		TEST_ASSERT_EQUAL_INT(_snret, snprintf(buf, _snlen, _format, _value)); \
+		TEST_ASSERT_EQUAL_STRING(_expect, buf); \
+	}
+
+
+/* group asserting pure usage of type specifiers */
+TEST_GROUP(fprintf_formatting);
+
+TEST_SETUP(fprintf_formatting)
+{
+	wr = fopen(stdpath, "w");
+	rd = fopen(stdpath, "r");
+}
+
+
+TEST_TEAR_DOWN(fprintf_formatting)
+{
+	fclose(wr);
+	fclose(rd);
+	remove(stdpath);
+}
+
+
+TEST(fprintf_formatting, format_int)
+{
+	/* integers */
+	ftest("%d", 123, "123");
+	ftest("%ld", (long)123, "123");
+	ftest("%i", 123, "123");
+	ftest("%o", 123, "173");
+	ftest("%x", 123, "7b");
+	ftest("%X", 123, "7B");
+}
+
+
+TEST(fprintf_formatting, format_float)
+{
+	/* floating point */
+	ftest("%f", 392.69, "392.690002");
```

Comment:
There is a problem with %f not detected by this test - %f should print double precision, but in our current implementation single precision (i.e. float) is used instead. You can test for that by attempting to print a value of a precision not provided by float but provided by double - result currently will not match

Reply:
As new issues were found in `printf()`-ing floats (https://github.com/phoenix-rtos/phoenix-rtos-project/issues/400) I have added checks for precision specific numbers which are marked as ignored. Trivial small numbers that you mentioned were left in place (392.69) for regression tests. Some future changes may even disturb currently working float printing, thus I am leaving it in place.